jax/_src/cudnn/fused_attention_stableHLO.py

```diff
@@ -0,0 +1,672 @@
+    value_dtype = dtypes.canonicalize_dtype(value.dtype)
+
+    return (
+        ShapedArray(
+            query.shape, query_dtype
+        ),  # grad query
+        ShapedArray(
+            key.shape, key_dtype
+        ),  # grad key
+        ShapedArray(
+            value.shape, value_dtype
+        ),  # part value
+    )
+
+def _dot_product_attention_fwd_cuda_lowering(ctx, query, key, value, bias, mask,
+    scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+    query_type = ir.RankedTensorType(query.type)
+    query_shape = query_type.shape
+    key_type = ir.RankedTensorType(key.type)
+    key_shape = key_type.shape
+    value_type = ir.RankedTensorType(value.type)
+    value_shape = value_type.shape
+    
+    batch, q_seq_len, num_heads, head_dim = query_shape
+    _, kv_seq_len, _, _ = key_shape
+
+    output_shape = (batch, num_heads, q_seq_len, head_dim)
+    output_layout = (3, 1, 2, 0)
+    output_transpose_perm = mlir.dense_int_array((0, 2, 1, 3))
+    activation_shape = (batch, num_heads, q_seq_len, kv_seq_len)
+    softmax_stat_shape = (batch, num_heads, q_seq_len)
+    scratch_shape = (0,)
+    scratch_type = ir.IntegerType.get_unsigned(8)
+    # get backend config
+    backend_config = create_dot_product_attention_backend_config(batch, num_heads, q_seq_len, kv_seq_len, query_type.element_type, scale, seed, dropout_rate, is_flash_attention, is_causal_mask, False)
+    # {Q, K, V, mask*, bias*}
+    # {output, scratch, activation*}
+    has_dropout = dropout_rate > 0
+    has_bias, has_mask = variadic_args
+    operands = [query, key, value]
+    if has_mask:
+        operands.append(mask)
+    if has_bias:
+        operands.append(bias)
+    # get custom call name
+    custom_call_name = get_custom_call_name(has_bias, has_mask, has_dropout, False)
+    # create output types and layouts
+    if is_flash_attention:
+        result_types = [
+            ir.RankedTensorType.get(output_shape, query_type.element_type),
+            ir.RankedTensorType.get(scratch_shape, scratch_type),
+            ir.RankedTensorType.get(softmax_stat_shape, ir.F32Type.get()),
+        ]
+        result_layouts = [output_layout] + default_layouts(scratch_shape, softmax_stat_shape)
+    else:
+        result_types = [
+            ir.RankedTensorType.get(output_shape, query_type.element_type),
+            ir.RankedTensorType.get(scratch_shape, scratch_type),
+            ir.RankedTensorType.get(activation_shape, query_type.element_type),
+        ]
+        result_layouts = [output_layout] + default_layouts(scratch_shape, activation_shape)
+    # create custom call here
+    out = custom_call(
+        custom_call_name,
+        result_types=result_types,
+        operands=operands,
+        backend_config=backend_config,
+        operand_layouts=default_layouts(*[ir.RankedTensorType(operand.type).shape for operand in operands]),
+        result_layouts=result_layouts,
+    )
+    # dropout scratch memory
```

Comment:
Nit: "dropout scratch memory" -> "drop scratch memory" (dropout is a bit confusing in context)